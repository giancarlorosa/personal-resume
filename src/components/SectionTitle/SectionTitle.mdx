import { Meta, Canvas, Controls } from '@storybook/blocks';
import * as SectionTitleStories from './SectionTitle.stories.jsx';

<Meta of={SectionTitleStories} />

# Section Title Component

The **Section Title** component was developed to serve the `H2` tag helping to name different sections of the page. Its implementation is very simple and just asks for a **string** as a parameter. Instead of using **props**, it was decided that using **children** could be more semantic and readable. The component receives a parameter called **fullWidth** that makes its underline element **100% width** of its container.

## Implementation

This component has a very simple implementation. Just inform a **string** as children.
**DO NOT USE NODE INSIDE CHILDREN PROP!**

If you want a **100% width**, just set the **fullWidth** parameter.

<Canvas of={SectionTitleStories.Implementation} />

<Controls of={SectionTitleStories.Implementation} />

## Multiple Lines

Here it's possible to test the multiple lines inside the component to check its line height.

<Canvas of={SectionTitleStories.MultipleLines} />

<Controls of={SectionTitleStories.MultipleLines} />

## Followed by a paragraph

Here it's possible to check the component margin bottom when it's followed by another element. The component container has a limited width to be possible to see multiple lines inside the container.

<Canvas of={SectionTitleStories.WithParagraph} />

<Controls of={SectionTitleStories.WithParagraph} />
