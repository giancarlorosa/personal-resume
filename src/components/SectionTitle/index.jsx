import PropTypes from 'prop-types';
import classNames from 'classnames';

function SectionTitle({ children, fullWidth }) {
  const sectionTitleClasses = classNames(
    'text-2xl',
    'text-dark-jungle-green',
    'font-semibold',
    'relative',
    'mb-7',
    { 'inline-block': fullWidth === false },
    'after:bg-mercury',
    'after:block',
    'after:content-[""]',
    'after:h-0.5',
    'after:mt-1',
    'after:w-14',
    'after:w-full',
    'before:bg-deep-sky-blue',
    'before:block',
    'before:content-[""]',
    'before:h-0.5',
    'before:mt-1',
    'before:w-8',
    'before:absolute',
    'before:bottom-0'
  );

  return (
    <div>
      <h2 className={sectionTitleClasses}>{children}</h2>
    </div>
  );
}

SectionTitle.propTypes = {
  children: PropTypes.string.isRequired,
  fullWidth: PropTypes.bool,
};

SectionTitle.defaultProps = {
  fullWidth: false,
};

export default SectionTitle;
