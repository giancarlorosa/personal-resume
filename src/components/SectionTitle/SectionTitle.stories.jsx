import SectionTitle from '.';

export default {
  title: 'Components/Section Title',
  component: SectionTitle,
};

const defaultArgTypes = {
  children: { control: 'text' },
  fullWidth: { control: 'boolean' },
};

export const Implementation = {
  args: {
    children: 'Section Title',
    fullWidth: false,
  },
  argTypes: defaultArgTypes,
};

export const MultipleLines = {
  args: {
    children: 'Section Title With Multiple Lines To Test the Line Height',
    fullWidth: false,
  },
  argTypes: defaultArgTypes,
  render: ({ children, fullWidth }) => (
    <div className="max-w-lg">
      <SectionTitle fullWidth={fullWidth}>{children}</SectionTitle>
    </div>
  ),
};
