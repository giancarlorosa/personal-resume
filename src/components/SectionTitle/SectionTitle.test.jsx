/* eslint-disable no-console */
import { render } from '@testing-library/react';
// TODO: Check why toHaveClass is imported but not recognized.
// eslint-disable-next-line no-unused-vars
import { toHaveClass } from '@testing-library/jest-dom';
import SectionTitle from '.';

describe('Section Title', () => {
  beforeAll(() => jest.spyOn(console, 'error').mockImplementation(() => {}));
  afterAll(() => console.error.mockRestore());
  afterEach(() => console.error.mockClear());

  it('should render the component', () => {
    const { getByText } = render(<SectionTitle>Section Title</SectionTitle>);
    const componentTitle = getByText('Section Title');
    expect(componentTitle).toBeTruthy();
  });

  it('should verify if component has a H2 tag', () => {
    const { getByRole } = render(<SectionTitle>Section Title</SectionTitle>);
    const componentTag = getByRole('heading', { level: 2 });
    expect(componentTag).toBeTruthy();
  });

  it('should have by default the display inline-block class when rendering the component', () => {
    const { getByText } = render(<SectionTitle>Section Title</SectionTitle>);
    const componentTitle = getByText('Section Title');
    expect(componentTitle).toHaveClass('inline-block');
  });

  it('should not have display inline-block class if fullWidth parameter is set', () => {
    const { getByText } = render(
      <SectionTitle fullWidth>Section Title</SectionTitle>
    );

    const componentTitle = getByText('Section Title');
    expect(componentTitle).not.toHaveClass('inline-block');
  });

  it.each([
    {
      errorLabel: 'children prop is not set',
      expectText: 'The prop `children` is marked as required in `SectionTitle`',
    },
    {
      children: <div>Section Title</div>,
      errorLabel: 'children prop is not a string',
      expectText:
        'Invalid prop `children` of type `object` supplied to `SectionTitle`',
    },
    {
      children: 'Section Title',
      fullWidth: 'full',
      errorLabel: 'fullWidth prop is not a boolean',
      expectText:
        'Invalid prop `fullWidth` of type `string` supplied to `SectionTitle`',
    },
  ])(
    'should throw an error when $errorLabel',
    ({ children, fullWidth, expectText }) => {
      render(<SectionTitle fullWidth={fullWidth}>{children}</SectionTitle>);
      expect(console.error).toHaveBeenCalled();
      expect(console.error.mock.calls[0][2]).toContain(expectText);
    }
  );
});
