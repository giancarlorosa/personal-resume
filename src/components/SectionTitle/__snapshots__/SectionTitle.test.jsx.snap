// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section Title should verify changes on component styles after setting fullWidth parameter 1`] = `"text-2xl text-dark-jungle-green font-semibold relative mb-7 after:bg-mercury after:block after:content-[""] after:h-0.5 after:mt-1 after:w-14 after:w-full before:bg-deep-sky-blue before:block before:content-[""] before:h-0.5 before:mt-1 before:w-8 before:absolute before:bottom-0"`;

exports[`Section Title should verify changes on component styles on its default state 1`] = `"text-2xl text-dark-jungle-green font-semibold relative mb-7 inline-block after:bg-mercury after:block after:content-[""] after:h-0.5 after:mt-1 after:w-14 after:w-full before:bg-deep-sky-blue before:block before:content-[""] before:h-0.5 before:mt-1 before:w-8 before:absolute before:bottom-0"`;
